### PR TITLE
[clang] Diagnose use of deprecated template alias

### LINKED ISSUE #18236: Support attribute deprecated for template alias
|  |  |
| --- | --- |
| Bugzilla Link | [17862](https://llvm.org/bz17862) |
| Version | trunk |
| OS | Linux |
| CC | @DougGregor |

## Extended Description 
I want to rename a templated class. To make the transition easier for the users, I'd like to keep the old class for one more version and mark it deprecated with the extensions from GCC / Clang (attribute deprecated). To avoid keeping an exact copy of the deprecated class, the use of template alias would be handy. Unfortunately Clang (3.3 and recent trunk) does not emit a warning (GCC does) for the following code in line 14:

```cpp
template <class T>
struct NewClassName
{
    // ...
};

template <class T> using OldClassNameUsing __attribute__ ((deprecated))
  = NewClassName<T>;

typedef NewClassName<int> OldClassNameTypedef __attribute__ ((deprecated));

int main()
{
  OldClassNameUsing<int> objectUsing;
  OldClassNameTypedef objectTypedef;

  return 0;
}
```

Clang used:
clang version 3.4 (194323)
Target: x86_64-unknown-linux-gnu
Thread model: posix

### PR DESCRIPTION
**Pull request by premanandrao**
Issue a warning diagnostic when a template alias with a deprecated attribute is used.

Fixes #18236.

### CODE REVIEW

**Comment by mizvekov** [comment]
For reference, see https://reviews.llvm.org/D136533 for similar expansion of diagnostics for uses of declarations, and how that lead to trouble in MacOS land.

I'd suggest building libc++ on MacOS, to be sure this won't cause regressions and a future revert.

I have a MacOS machine now, so I should be able to try to revive that patch at some point as well.

**Comment by premanandrao** [comment]
> For reference, see https://reviews.llvm.org/D136533 for similar expansion of diagnostics for uses of declarations, and how that lead to trouble in MacOS land.
> I'd suggest building libc++ on MacOS, to be sure this won't cause regressions and a future revert.

Thanks for pointing out the other PR.  I don't have a MacOS system handy right now, but I can get to one after the weekend.

> I have a MacOS machine now, so I should be able to try to revive that patch at some point as well.

Sounds good.

**Comment by cor3ntin** [comment]
@ldionne

**Comment by ldionne** [comment]
I just ran the bootstrapping build with this patch applied and ran `check-runtimes`, and nothing failed. This might not be sufficient (perhaps I missed something important), but at first glance it doesn't seem to break anything on our side.

**Comment by premanandrao** [comment]
> I just ran the bootstrapping build with this patch applied and ran `check-runtimes`, and nothing failed. This might not be sufficient (perhaps I missed something important), but at first glance it doesn't seem to break anything on our side.

Thank you SO much!  

Would you mind sharing (pointers to) instructions on how you built this?  I was running into errors about __remove_cv_t (and others) at my end.

**Comment by ldionne** [comment]
@premanandrao  There are instructions here: https://libcxx.llvm.org/BuildingLibcxx.html#bootstrapping-build

You can also use e.g. `run-buildbot` for this job: https://github.com/llvm/llvm-project/blob/c7961538ff5c73ad03cbf2470e56cdc10cedc83b/libcxx/utils/ci/run-buildbot#L366
You will probably need to modify a few things like maybe remove the mentions of `ccache`.

File: clang/test/SemaTemplate/alias-template-deprecated.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// template triggers a warning diagnostic when it is used.
+
+template <typename T>
+struct NoAttr {
+  void foo() {}
+};
+
+// expected-note@+2 5{{'UsingWithAttr' has been explicitly marked deprecated here}}
+template <typename T>
+using UsingWithAttr __attribute__((deprecated)) = NoAttr<T>;
```
Comment (tahonermann) [inline clang/test/SemaTemplate/alias-template-deprecated.cpp:13]: I recommend adding a test that uses the C++11 `[[deprecated]]` attribute syntax.

Comment (premanandrao) [reply clang/test/SemaTemplate/alias-template-deprecated.cpp:13]: Thank you @tahonermann!  Done.

File: clang/test/SemaTemplate/alias-template-deprecated.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  // expected-warning@+1 {{'UsingTDWithAttr' is deprecated}}
+  UsingTDWithAttr objUTDWA;
+
+  // expected-warning@+2 {{'UsingWithCPPAttr' is deprecated}}
+  // expected-note@+1 {{in instantiation of template type alias 'UsingWithCPPAttr' requested here}}
+  UsingWithCPPAttr<int> objUsingWCPPA;
+
+  // expected-warning@+1 {{'UsingInstWithCPPAttr' is deprecated: Do not use this}}
+  UsingInstWithCPPAttr objUICPPWA;
```
Comment (AaronBallman) [inline clang/test/SemaTemplate/alias-template-deprecated.cpp:71]: Some more test cases I'd like to see:
```
NoAttr<UsingWithAttr<int>> s; // diag

using Foo [[deprecated]] = int;
using X = UsingWithAttr<Foo>; // two diagnostics, one for Foo and one for UsingWithAttr
```

**Review by AaronBallman — COMMENTED** [review]
Thank you for this, the changes should also come with a release note in clang/docs/ReleaseNotes.rst so users know about the improvement. Changes generally LG, but I did have some test suggestions.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Review by mizvekov — APPROVED** [review]
LGTM as well.

**Pull request merged**